### ISSUE TITLE
(3d) isoform overlap rules seem to be inconsistently applied

### ISSUE THREAD

**Issue by nathandunn** [GMOD/Apollo#498]
I would expect if you drag up a transcript where part of the CDS overlapped . . . even if only on one exon, that it is a valid isoform.  Not in WA2 . . . not in WA1.   This data is Group 1.1 honeybee.   Fixed so same region . . . same error / encoding. Different naming schema only.  

WA2:
<img width="329" alt="screen shot 2015-07-23 at 3 04 35 pm" src="https://cloud.githubusercontent.com/assets/751274/8863283/29816d60-314c-11e5-8aac-7da906c4fa4c.png">

WA1:

<img width="400" alt="screen shot 2015-07-23 at 2 57 46 pm" src="https://cloud.githubusercontent.com/assets/751274/8863182/3a99c81e-314b-11e5-9139-5b76e986cf93.png">

**Comment by nathandunn** [comment]
@selewis   Sanity check requested when you have a moment.

**Comment by nathandunn** [comment]
This happens only for negatively stranded isoforms.  

WA1 works correctly.

**Comment by selewis** [comment]
I haven't checked this out yet, but have a question before I do.

It's at least one exon from each transcript overlap AND

*they are in the same reading frame*
If it is just the UTR (of either one) then different genes
If it is different reading frames then different genes

Is this the case from your software perspective? Was it implemented with
this intent?

-S

**Comment by selewis** [comment]
p.s. please send link for where to test this.

**Comment by nathandunn** [comment]
It ends up that it was a bug only related to the negative strand.  Works perfectly on positive strand.   I just need to fix it / figure out where it was introduced. 

Nathan

**Comment by nathandunn** [comment]
I think that this bears a bit of discussion and probably can be done as part of #73 . . .

**Comment by nathandunn** [comment]
Other genes to check:

GB42148:  adding in any order yields this:

![screen shot 2015-07-24 at 9 41 38 am](https://cloud.githubusercontent.com/assets/751274/8879738/3e682e7a-31e8-11e5-8941-114f56f33958.png)

I get the expected correct behavior with GB42182 however (other strand).

**Comment by nathandunn** [comment]
So, was forgetting to check the reading frames . . . however, when I look over some other examples, I think that there is still a problem:

![screen shot 2015-07-24 at 2 05 07 pm](https://cloud.githubusercontent.com/assets/751274/8884627/11a773f8-320d-11e5-8159-11ecb63339c1.png)

**Comment by nathandunn** [comment]
I would expect ```GB40772-RA-00002``` and ```au9.g284.t1-00002```

**Comment by nathandunn** [comment]
@selewis   
So, the reason these don't work is that, though the CDS is in-phase, the overlap function is comparing the upstream exon location to calculate phase.  

This is identical in WA2 and WA1.  I believe identically wrong.  

The current WA1 adopted algorithm for determining if overlapping isoform:
- get CDS1 and CDS2 for transcript 1 and transcript 2, respectively and determine if they overlap
- determine if at least one exon between the two transcripts (whether or not in the CDS) overlaps and has the phase as calculated by the upstream exon start. 

Proposed WA2 algorithm:
- compare strand of the transcripts 
- get CDS1 and CDS2 for transcript 1 and transcript 2, respectively and determine if they overlap or are in phase 
- for every exon from within the CDS1 and CDS2 find at least one overlap.

**Comment by nathandunn** [comment]
In the cases sited above (https://github.com/GMOD/Apollo/issues/498#issuecomment-124723845) the proteins are identical.

**Comment by nathandunn** [comment]
FYI @monicacecilia  (when you have time) let me know if I am off-base on this, or if I am correct.

**Comment by selewis** [comment]
comments on WA2 proposal

On Fri, Jul 24, 2015 at 3:35 PM, Nathan Dunn <notifications@github.com>
wrote:

> @selewis <https://github.com/selewis>
>
> So, the reason these don't work is that, though the CDS is in-phase, the
> overlap function is comparing the upstream exon location to calculate
> phase.
>
> This is identical in WA2 and WA1. I believe identically wrong.
>
> The current WA1 adopted algorithm for determining if overlapping isoform:
>
>    - get CDS1 and CDS2 for transcript 1 and transcript 2, respectively
>    and determine if they overlap
>    - determine if at least one exon between the two transcripts (whether
>    or not in the CDS) overlaps and has the phase as calculated by the upstream
>    exon start.
>
> Proposed WA2 algorithm:
>
>    - compare strand of the transcripts
>
>
>    - get CDS1 and CDS2 for transcript 1 and transcript 2, respectively
>    and determine if they overlap or are in phase
>
> boolean different_genes = (transcript1 strand == transcript2 strand)
TSS1 = get translation-start-site (TSS) of transcript 1
TSS2 = get translation-start-site of transcript2
if (!different_genes && CDS1 overlaps CDS2 && phase of TSS1 == phase of
TSS2)
    different_genes = false;

In other words, you don't have to look at phase except for where
translation starts (and the same strand, and the CDS regions overlap)

**Comment by nathandunn** [comment]
Thanks.  It sounds like my proposal is correct then.  Just surprised that this was in the old code as its really easy to reproduce.    We'll need to do this with the overlapping function as well.  

Nathan

**Comment by selewis** [comment]
Right,

The only thing I was noting is that you don't have to check the phase of
each exon in the CDS, just the TSS genomic phase has to be the same (and
you said 'or')

**Comment by nathandunn** [comment]
Yes this makes sense.   

Yes the CDS must overlap *AND* the transcription start phase must be the same.  

What I meant, but not what I said.  I’ll fix the text. 

Thanks,

Nathan

**Comment by nathandunn** [comment]
From a brief conversation with @monicacecilia   . . . everything in here is correct, with the caveat that Exons will almost always overlap exactly.    In conjunction with #73 this might end up being a non-issue, but it should still be done correctly. 

@selewis   This is the code that is "wrong" . . .coming straight from WA1:

https://github.com/GMOD/Apollo/blob/master/grails-app/services/org/bbop/apollo/OverlapperService.groovy#L69

```
  if (exon1FeatureLocation.strand.equals(Strand.POSITIVE.value) && exon1FeatureLocation.getFmin() % 3 == exon2FeatureLocation.getFmin() % 3) {
                        return true;
                    }
                    else if (exon1.strand.equals(Strand.NEGATIVE.value) && exon1FeatureLocation.getFmax() % 3 == exon2FeatureLocation.getFmax() % 3) {
                        return true;
                    }
```

**Comment by cmdcolin** [comment]
I would recommend revisiting this

Many instances of transcripts that should be part of the same gene are not merged due to bad overlapping strategies

WA1.0 probably had it wrong, but WA2.0 is not very good now either because practically everything, even transcripts that are 100% overlapping CDS and slightly different utr, are counted as a separate gene

**Comment by monicacecilia** [comment]
Yes, it's currently tagged for 2.0.1, on the list to review again next.

One quick note: I don't understand what you mean when describing "transcripts that should be part of the same gene are not merged". Are you expecting automatic behaviour that merges transcripts? The function "to merge" means bringing two or more transcripts (adjacent or not; same strand) together, and it only happens when the curator executes the function. It may be the case that you are describing a different scenario here and that is why I cannot understand the line. - it would be helpful if you could please add a figure (screen captures or a diagram) to further explain. Thank you! :)

cheers,

**Comment by cmdcolin** [comment]
Merge was probably not the best choice of works. I meant that they should belong to the same gene model as alternative isoforms.

**Comment by nathandunn** [comment]
What @cmdcolin  said.  I'll fix it.

**Comment by nathandunn** [comment]
I had the same understanding as @cmdcolin . . . I agree with that clarification.

**Comment by monicacecilia** [comment]
10-4.

**Comment by childers** [comment]
Hi all,
I just received this email from one of our high volume annotators, and while it is related to isoforms, I'm not sure if it is related.  Please feel free to split off to another issue if this is the case. I am following up for more specific information, and hope to have some screenshots or links for you soon.

Chris, for some time now I've experienced a strange problem with Web
Apollo where if I try to establish multiple gene models in the same
region, it insists on putting them on top of each other. This makes some
things difficult to do, like alternatively-spliced models. This is using
Chrome. It started a couple of months ago, but I did not bother you with
it then because I was not building alternatively-spliced models.

**Comment by nathandunn** [comment]
So . . the general rule we are trying to enforce right now (correct me if I’m wrong anyone) is that if isoforms have overlapping (and in-phase / same-strand) CDS regions . . they must belong to the same gene model.  

It *sounds* like they have two problems:
1 - they are trying to establish multiple gene models (not isoforms) in the same region.  This is allowed now . . but not for long (once this bug is finished). 
2 - it is actually altering the exon (or other boundaries) off of what they were trying to set 

Maybe I’m interpreting “same region” as “overlapping region”, though.   Can you provide more clarification?  

Nathan

**Comment by cmdcolin** [comment]
They might actually just have the "collapse" setting on with the annotator track?

**Comment by childers** [comment]
I had just suggested that as well.  I had forgotten that was an option on the User-created Annotations track until I was trying to recreate this issue.

**Comment by nathandunn** [comment]
I think that @deepakunni3  is still working on this.

**Comment by monicacecilia** [comment]
I think this will be solved when #73 is solved. Let's leave them linked and close at the same time --- I think that @nathandunn encountered similar issues to the ones I noted on 2015-10-26

**Comment by monicacecilia** [comment]
I'm assigning @deepakunni3 since this is linked to #73, which is assigned to him.

**Comment by monicacecilia** [comment]
Still pending on #73, which is not yet ready.

**Comment by monicacecilia** [comment]
#73 is now closed. This can be closed - working as expected. :dancer:
